### ISSUE TITLE
"method with an Action as a single argument" does not help implementing the Configurable interface

### ISSUE THREAD

**Issue by KengoTODA** [gradle/gradle#19073]
I’m trying to remove dependencies on `ConfigureUtil.configureSelf()` in my Gradle plugin, which is used to implement the [`Configurable` interface](https://docs.gradle.org/7.3/javadoc/org/gradle/util/Configurable.html). Here is a pseudoclass:

```kotlin
abstract class ConfigurableClass : Configurable<ConfigurableClass> {
    public fun configure(closure: Closure): ConfigurableClass {
        return ConfigureUtil.configureSelf(closure, this)
    }
}
```

The [latest Javadoc](https://docs.gradle.org/7.3/javadoc/org/gradle/util/ConfigureUtil.html) says I just need a “method with an Action as a single argument” so I created a method as follows:

```kotlin
// kotlin
abstract class ConfigurableClass : Configurable<ConfigurableClass> {
    public fun configure(action: Action<ConfigurableClass>): ConfigurableClass {
        action.execute(this)
        return this
    }
}
```

### Expected Behavior

Gradle will create a `configure(Closure)` method based on the provided `configure(Action)` method, so we can safely remove the dependency on the deprecated `ConfigureUtil` class.

### Current Behavior

It seems not working to satisfy the `configure()` method required by the `Configurable` interface.
On runtime Gradle throws exception and its error message says like below:

```
// * What went wrong:
// Could not create an instance of type Build_gradle$ConfigurableClass.
// > Could not generate a decorated class for type Build_gradle.ConfigurableClass.
//    > Cannot have abstract method Configurable.configure().
```

### Context

I am working for [this Report implementation](https://github.com/spotbugs/spotbugs-gradle-plugin/blob/5ff6f32bb39a8acdc1db2c64963eb3b355e33a9d/src/main/groovy/com/github/spotbugs/snom/SpotBugsReport.java#L114-L118). To support both Gradle v7 and v8, I want to remove usage of deprecated API with proper one.
Note that [the `Report` interface extends the `Configurable` interface](https://docs.gradle.org/7.3/javadoc/org/gradle/api/reporting/Report.html). So I cannot stop implementing the `Configurable` interface.

I searched official documents and found no migration guide. I posted at Gradle Slack and [Gradle Forum](https://discuss.gradle.org/t/how-to-replace-the-usage-of-deprecated-configureutil-configureself/41455?u=kengotoda) but no help I found.

### Steps to Reproduce 

Please clone https://github.com/KengoTODA/mcve-configurable and run `./gradlew build`.

### Your Environment

Build scan URL: https://gradle.com/s/c3xveyiuw4vmw

**Comment by ljacomet** [comment]
You should no longer implement the `Configurable` interface. Instead you should have an `Action` taking method on your extension and when instantiating it through Gradle, it will be decorated to have all the necessary bells and whistles for the DSL usage.

**Comment by KengoTODA** [comment]
@ljacomet thanks for your notice. let me confirm one thing:

> You should no longer implement the Configurable interface

I need to implement it because of the following reason:

> Note that the Report interface extends the Configurable interface. So I cannot stop implementing the Configurable interface.

Can I understand that the `Report` interface will be changed and it will not extend the `Configurable` interface?

**Comment by ljacomet** [comment]
🤔 I missed that ...
Do you use that method at all explicitly?

From the plugin documentation, I believe this method does not get exercised, the nesting of blocks is handled by the `NamedDomainObjectContainer`. So I would try making this method an error and instead have a dedicated action taking method, which will result in a properly decorated type for the DSL.

**Comment by KengoTODA** [comment]
> Do you use that method at all explicitly?

Yes for now at https://github.com/spotbugs/spotbugs-gradle-plugin/blob/932b215d996bce082b0dfab964fb8941332c471a/src/main/groovy/com/github/spotbugs/snom/SpotBugsReport.java

I made this abstract class to add several methods over the `SingleFileReport` interface and the `CustomizableHtmlReport` interface.


> the nesting of blocks is handled by the NamedDomainObjectContainer

It is true even in our project. Here is the related code to create `Report` instances:

https://github.com/spotbugs/spotbugs-gradle-plugin/blob/932b215d996bce082b0dfab964fb8941332c471a/src/main/groovy/com/github/spotbugs/snom/SpotBugsTask.groovy#L316-L331

**Comment by KengoTODA** [comment]
Ah maybe I found a solution; instead of an abstract class, use the `CustomizableHtmlReport` then its instance can be generated by `NamedDomainObjectContainer`. We cannot add new methods, but can find other way to generate different command line for each report type.

**Comment by ljacomet** [comment]
Sorry, I might not have been fully clear.

I would keep your `Report` classes hierarchy as-is, throw in `configure` instead of using `ConfigureUtil` and instead - if actually needed - add a dedicated `Action` taking method.

But from what I see in the implementation of `NamedDomainObjectContainer`, there is no strong requirement for using `Configurable`.

**Comment by KengoTODA** [comment]
> But from what I see in the implementation of NamedDomainObjectContainer, there is no strong requirement for using Configurable.

What I want to achieve is extending the `CustomizableHtmlReport` interface, to provide several utilities like `reports { html { setStylesheet(notTextResourceButString) } }`. To achieve this, I create an abstract class and implement it for each report type like `XmlReport`, `HtmlReport`, etc.

The `CustomizableHtmlReport` interface extends the `Configurable` interface, so I need to implement the `configure(Closure)` method, or Gradle reports a `"Cannot have abstract method"` error on runtime.

**Comment by ljacomet** [comment]
Ah right, and because that interface is part of the type, Gradle will attempt to use it.

Let me re-open this, we may need to revise getting rid of `ConfigureUtil` in Gradle 8.

**Comment by KengoTODA** [comment]
Thanks for your cooperation, and sorry that I made too simple MCVE 😓
Probably I should make a MCVE based on the Report API.

Now I feel that the root cause is the `Report` API is not enough extensible for plugin developers.

Implementations of the `Report` interface like [SimpleReport](https://github.com/gradle/gradle/blob/ae3758aa7698196c03e7290116ea4b13388986c0/subprojects/reporting/src/main/java/org/gradle/api/reporting/internal/SimpleReport.java) and [TaskGeneratedReport](https://github.com/gradle/gradle/blob/ae3758aa7698196c03e7290116ea4b13388986c0/subprojects/reporting/src/main/java/org/gradle/api/reporting/internal/TaskGeneratedReport.java) are internal, so plugin developers need to implement the `Report` interface from scratch, so need to depends on `ConfigurableUtil` just like the `SimpleReport` class:

https://github.com/gradle/gradle/blob/ae3758aa7698196c03e7290116ea4b13388986c0/subprojects/reporting/src/main/java/org/gradle/api/reporting/internal/SimpleReport.java#L87-L90

**Comment by jeffalder** [comment]
With 8.0 released this is coming to the fore. @ljacomet any updates on this consideration?

**Comment by desoliture** [comment]
Hello, is there any updates on this?

**Comment by ov7a** [comment]
Sorry for the late reply.

Sorry that you're having trouble with Gradle!

The issue is in the backlog of the relevant team, but this area of Gradle is currently not a focus one, so it might take a while before a fix is made.

***

Our available resources are severely limited, unfortunately. I can't give you any ETA on any progress for this.

**Comment by Goooler** [comment]
We can add `org.gradle.api.Project#configureSelf(java.lang.Object, groovy.lang.Closure)` to get rid of `org.gradle.util.ConfigureUtil#configureSelf(groovy.lang.Closure, T)`.

**Comment by desoliture** [comment]
@Goooler If Gradle trying to deprecate Closures, maybe the Action instead of Closure be a good choice, isnit?

**Comment by Goooler** [comment]
Yeah, I tried to replace [com.github.spotbugs.snom.SpotBugsReport#configure](https://github.com/spotbugs/spotbugs-gradle-plugin/blob/fda1755cd6be5a6101c576904ea314b35d6b74ff/src/main/groovy/com/github/spotbugs/snom/SpotBugsReport.java#L119-L123) with the Action implementation, but it extends `org.gradle.api.reporting.Report` which is a subclass of `Configurable`, must implement `configure` method; and if I just want to replace the usage of `ConfigureUtil.configureSelf`, there is no fallback in `Project`, `Project. configure` can't be used here cause `SpotBugsReport` is a subclass of `Configurable`, which will result in a recursive call.

**Comment by Goooler** [comment]
Workaround:

```java
public class Foo implements Configurable<Foo> {

  @Override
  public Foo configure(Closure closure) {
    return configure(
        report -> {
          closure.setDelegate(report);
          closure.call(report);
        });
  }

  public Foo configure(Action<? super Foo> action) {
    action.execute(this);
    return this;
  }
}
```
